builder/codebuild/codebuild.go

```diff
@@ -0,0 +1,353 @@
+	Dry         bool
+	DockerCfg   string
+}
+
+// Builder is a builder.Builder implementation that builds Docker images using
+// CodeBuild.
+type Builder struct {
+	// BuildSpec is a template.Template that will be executed to generate a
+	// buildspec.yml file for CodeBuild. It's given a builder.BuildOptions
+	// as data, and is expected to return a valid buildspec.yml that will:
+	//
+	// 1. Build the Docker Image.
+	// 2. Push it to a registry.
+	BuildSpec *template.Template
+
+	// Role that will be provided to the CodeBuild project.
+	ServiceRole string
+
+	// Dry specifies whether "Dry" run mode is enabled. When true, the
+	// CodeBuild build won't push to the Docker registry.
+	Dry bool
+
+	// DockerCfg can be provided to pass credentials to Docker. This
+	// should point to an SSM parameter containing a valid .dockercfg.
+	DockerCfg string
+
+	// Function called to determine the CodeBuild project name, given a
+	// repository.
+	ProjectName func(builder.BuildOptions) string
+
+	codebuild codebuildClient
+
+	// This gets called when we need an io.Reader for a CloudWatch log
+	// stream.
+	logReader func(group, stream string) (io.ReadCloser, error)
+}
+
+func NewBuilder(config client.ConfigProvider) *Builder {
+	cloudwatchlogs := cloudwatchlogs.New(config)
+
+	// Use the cloudwatch package to get an io.Reader for a cloudwatch log
+	// stream.
+	logReader := func(group, stream string) (io.ReadCloser, error) {
+		return cloudwatch.NewReader(group, stream, cloudwatchlogs), nil
+	}
+
+	return &Builder{
+		BuildSpec:   template.Must(template.New("buildspec.yml").Parse(DefaultBuildspec)),
+		ProjectName: ProjectName,
+		DockerCfg:   "conveyor.dockercfg",
+		codebuild:   codebuild.New(config),
+		logReader:   logReader,
+	}
+}
+
+// Build executes the docker image.
+func (b *Builder) Build(ctx context.Context, w io.Writer, opts builder.BuildOptions) (image string, err error) {
+	projectName := b.ProjectName(opts)
+	image = fmt.Sprintf("%s:%s", opts.Repository, opts.Sha)
+	err = b.build(ctx, w, BuildOptions{
+		BuildOptions: opts,
+		ProjectName:  projectName,
+		Dry:          b.Dry,
+		DockerCfg:    b.DockerCfg,
+	})
+	return
+}
+
+func (b *Builder) build(ctx context.Context, w io.Writer, opts BuildOptions) error {
+	resp, err := b.startBuild(ctx, opts)
+	if err != nil {
+		// if the project does not exist, create it.
+		if awsErr, ok := err.(awserr.Error); ok && awsErr.Code() == "ResourceNotFoundException" {
+			if _, err := b.createProject(ctx, opts); err != nil {
+				return err
+			}
+
+			resp, err = b.startBuild(ctx, opts)
+			if err != nil {
+				return err
+			}
+		} else {
+			return fmt.Errorf("starting build for %s: %v", opts.Repository, err)
+		}
+	}
+
+	buildID := resp.Build.Id
+
+	done := make(chan error)
+	go func() {
+		done <- b.tailBuild(ctx, w, buildID)
+	}()
+
+	select {
+	case <-ctx.Done():
+		// Try cancelling the build. We don't really care about the
+		// error here.
+		b.codebuild.StopBuild(&codebuild.StopBuildInput{
```

Comment:
If this failed, you might have a zombie running build, but only until it hits the codebuild timeout (60 minutes), so this is just kinda a nice to have.